File: src/kavach/api/models/audit.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel


class AuditMetricResponse(BaseModel):
    label: str
    value: int
    description: str | None = None


class AuditFilterOptionsResponse(BaseModel):
    statuses: list[str]
    tool_names: list[str]
    actor_ids: list[str]
    resource_types: list[str]
    operation_types: list[str]


class AuditListItemResponse(BaseModel):
    audit_id: str
    request_id: str
    correlation_id: str
    tool_name: str
    operation_type: str
    resource_type: str
    resource_id: str | None = None
    actor_id: str
    status: str
    dry_run: bool
    started_at: datetime
    completed_at: datetime | None = None
    duration_ms: float | None = None
    interrupted: bool
    interrupted_reason: str | None = None
    reason: str
    job_id: str | None = None
    job_status: str | None = None
    result_reference: str | None = None
    error_code: str | None = None


class AuditLinkedJobResponse(BaseModel):
    job_id: str
    job_type: str
    status: str
    submitted_by: str
    updated_at: datetime


class AuditPageResponse(BaseModel):
    summary: list[AuditMetricResponse]
    filters: AuditFilterOptionsResponse
    records: list[AuditListItemResponse]
    total: int
    limit: int
    offset: int


class AuditDetailResponse(BaseModel):
    audit_id: str
    request_id: str
    correlation_id: str
    tool_name: str
    tool_version: str
    actor_id: str
    actor_type: str
    agent_name: str | None = None
    agent_session_id: str | None = None
    client_name: str | None = None
    client_version: str | None = None
    idempotency_key: str
    operation_type: str
    resource_type: str
    resource_id: str | None = None
    request_hash: str
    request_summary: dict[str, Any]
    resolved_versions: dict[str, Any]
    reason: str
    dry_run: bool
    status: str
    job_id: str | None = None
    result_reference: str | None = None
    error_code: str | None = None
    error_message: str | None = None
    started_at: datetime
    completed_at: datetime | None = None
    duration_ms: float | None = None
    metadata: dict[str, Any]
    interrupted: bool
    interrupted_reason: str | None = None
    linked_job: AuditLinkedJobResponse | None = None
    related_records: list[AuditListItemResponse]


class InvocationAuditListItemResponse(BaseModel):
    invocation_id: str
    request_id: str
    correlation_id: str
    tool_name: str
    tool_version: str
    actor_id: str | None = None
    client_id: str | None = None
    status: str
    authorization_decision: str
    request_hash: str
    response_classification: str | None = None
    response_hash: str | None = None
    error_category: str | None = None
    started_at: datetime
    completed_at: datetime | None = None
    duration_ms: float | None = None


class InvocationAuditPageResponse(BaseModel):
    records: list[InvocationAuditListItemResponse]
    total: int
    limit: int
    offset: int

File: console/types/audit.ts
export type AuditStatus =
  | "STARTED"
  | "SUCCEEDED"
  | "FAILED"
  | "DRY_RUN";

export type AuditMetricDto = {
  label: string;
  value: number;
  description: string | null;
};

export type AuditFilterOptionsDto = {
  statuses: string[];
  tool_names: string[];
  actor_ids: string[];
  resource_types: string[];
  operation_types: string[];
};

export type AuditListItemDto = {
  audit_id: string;
  request_id: string;
  correlation_id: string;
  tool_name: string;
  operation_type: string;
  resource_type: string;
  resource_id: string | null;
  actor_id: string;
  status: AuditStatus;
  dry_run: boolean;
  started_at: string;
  completed_at: string | null;
  duration_ms: number | null;
  interrupted: boolean;
  interrupted_reason: string | null;
  reason: string;
  job_id: string | null;
  job_status: string | null;
  result_reference: string | null;
  error_code: string | null;
};

export type AuditLinkedJobDto = {
  job_id: string;
  job_type: string;
  status: string;
  submitted_by: string;
  updated_at: string;
};

export type AuditPageDto = {
  summary: AuditMetricDto[];
  filters: AuditFilterOptionsDto;
  records: AuditListItemDto[];
  total: number;
  limit: number;
  offset: number;
};

export type AuditDetailDto = {
  audit_id: string;
  request_id: string;
  correlation_id: string;
  tool_name: string;
  tool_version: string;
  actor_id: string;
  actor_type: string;
  agent_name: string | null;
  agent_session_id: string | null;
  client_name: string | null;
  client_version: string | null;
  idempotency_key: string;
  operation_type: string;
  resource_type: string;
  resource_id: string | null;
  request_hash: string;
  request_summary: Record<string, unknown>;
  resolved_versions: Record<string, unknown>;
  reason: string;
  dry_run: boolean;
  status: AuditStatus;
  job_id: string | null;
  result_reference: string | null;
  error_code: string | null;
  error_message: string | null;
  started_at: string;
  completed_at: string | null;
  duration_ms: number | null;
  metadata: Record<string, unknown>;
  interrupted: boolean;
  interrupted_reason: string | null;
  linked_job: AuditLinkedJobDto | null;
  related_records: AuditListItemDto[];
};

export type AuditMetric = {
  label: string;
  value: number;
  description: string | null;
};

export type AuditFilterOptions = {
  statuses: string[];
  toolNames: string[];
  actorIds: string[];
  resourceTypes: string[];
  operationTypes: string[];
};

export type AuditListItem = {
  auditId: string;
  requestId: string;
  correlationId: string;
  toolName: string;
  operationType: string;
  resourceType: string;
  resourceId: string | null;
  actorId: string;
  status: AuditStatus;
  dryRun: boolean;
  startedAt: string;
  completedAt: string | null;
  durationMs: number | null;
  interrupted: boolean;
  interruptedReason: string | null;
  reason: string;
  jobId: string | null;
  jobStatus: string | null;
  resultReference: string | null;
  errorCode: string | null;
};

export type AuditLinkedJob = {
  jobId: string;
  jobType: string;
  status: string;
  submittedBy: string;
  updatedAt: string;
};

export type AuditPage = {
  summary: AuditMetric[];
  filters: AuditFilterOptions;
  records: AuditListItem[];
  total: number;
  limit: number;
  offset: number;
};

export type AuditDetail = {
  auditId: string;
  requestId: string;
  correlationId: string;
  toolName: string;
  toolVersion: string;
  actorId: string;
  actorType: string;
  agentName: string | null;
  agentSessionId: string | null;
  clientName: string | null;
  clientVersion: string | null;
  idempotencyKey: string;
  operationType: string;
  resourceType: string;
  resourceId: string | null;
  requestHash: string;
  requestSummary: Record<string, unknown>;
  resolvedVersions: Record<string, unknown>;
  reason: string;
  dryRun: boolean;
  status: AuditStatus;
  jobId: string | null;
  resultReference: string | null;
  errorCode: string | null;
  errorMessage: string | null;
  startedAt: string;
  completedAt: string | null;
  durationMs: number | null;
  metadata: Record<string, unknown>;
  interrupted: boolean;
  interruptedReason: string | null;
  linkedJob: AuditLinkedJob | null;
  relatedRecords: AuditListItem[];
};
